pyserver/main.py

```diff
@@ -0,0 +1,362 @@
+      "u/2/in_tok": TK_2_IN,
+      "u/2/out_tok" : TK_2_OUT,
+      "rows_to_claims" : wandb.Table(
+                           data=comms_to_claims_html,
+                           columns = ["comments", "claims"])
+    })
+ 
+  net_usage = {"total_tokens" : TK_2_TOT,
+               "prompt_tokens" : TK_2_IN,
+               "completion_tokens" : TK_2_OUT}
+
+
+  return {"claims_tree" : node_counts, "usage" : net_usage}
+
+def dedup_claims(claims:list)-> dict:
+  """
+  Given a list of claims for a given subtopic, identify which ones are near-duplicates
+  """
+  client = OpenAI()
+
+  # add claims with enumerated ids (relative to this subtopic only)
+  full_prompt = config.CLAIM_DEDUP_PROMPT
+  for i, orig_claim in enumerate(claims):
+    full_prompt += "\nclaimId"+str(i)+ ": " + orig_claim
+
+  response = client.chat.completions.create(
+    model = config.MODEL,
+    messages = [
+      {
+        "role": "system",
+        "content": config.SYSTEM_PROMPT
+      },
+      {
+            "role": "user",
+            "content": full_prompt
+      }
+    ],
+    temperature = 0.0,
+    response_format = {"type": "json_object"}
+  )
+  deduped_claims = response.choices[0].message.content
+  return {"dedup_claims" : json.loads(deduped_claims), "usage" : response.usage}
+
+#####################################
+# Step 3: Sort & deduplicate claims #
+#-----------------------------------#
+@app.put("/sort_claims_tree/")
+def sort_claims_tree(claims_tree:ClaimTree, sort_type : str = "deduplicate", log_to_wandb: bool = False)-> dict:
+  #TODO: do we ever want to deduplicate without sorting? or sort without deduplicating?
+  TK_IN = 0
+  TK_OUT = 0
+  TK_TOT = 0
+  
+  dupe_counts = {}
+  dupe_logs = []
+  nested_claims = {}
+
+  if sort_type == "deduplicate": 
+    for topic, topic_data in claims_tree.tree.items():
+      for subtopic, subtopic_data in topic_data["subtopics"].items():
+        # no need to deduplicate single claims
+        if subtopic_data["total"] > 1:
+          response = dedup_claims(subtopic_data["claims"])
+          deduped_claims = response["dedup_claims"]
+          usage = response["usage"]
+
+          # check for duplicates
+          has_dupes = False
+          if "nesting" in deduped_claims:
+            for claim_key, claim_vals in deduped_claims["nesting"].items():
+              if len(claim_vals) > 0:
+                has_dupes = True
+                # extract relative index
+                claim_id = int(claim_key.split("Id")[1])
+                dupe_keys = [int(dupe_claim_key.split("Id")[1]) for dupe_claim_key in claim_vals]
+                dupe_counts[subtopic_data["claims"][claim_id]] = dupe_keys
+
+          # for logging to wandb
+          if log_to_wandb:
+            dupe_logs.append(["\n".join(subtopic_data["claims"]), json.dumps(deduped_claims, indent=1)])
+
+          if has_dupes:
+            nested_claims[subtopic] = {"dupes" : deduped_claims, "og" : subtopic_data["claims"]}
```

Comment:
I'm a little brain foggy and not sure I'm following the logic here right, but does this actually produce exactly the same schema we're currently sending to the client?

might be helpful at this point to explicitly describe the shape of the data blob the pipeline returns -- will also help keep things in sync with the client going forward (e.g. after Brandon's updates to fix the missing-quote-attribution issue)

Reply:
yes this is only approximate for now — will test!